docs/cpp/source/notes/tensor_cuda_stream.rst

```diff
@@ -0,0 +1,173 @@
+
+.. tip::
+
+  You can request a stream from the high priority pool by setting isHighPriority to true, or a stream for a specific device
+  by setting device index (defaulting to the current CUDA stream's device index).
+
+2. Get the default CUDA stream, for the passed CUDA device, or for the current device if no device index is passed.
+
+.. code-block:: cpp
+  
+  CUDAStream getDefaultCUDAStream(DeviceIndex device_index = -1);
+
+.. tip::
+  
+  The default stream is where most computation occurs when you aren't explicitly using streams.
+
+3. Get the current CUDA stream, for the CUDA device with index ``device_index``, or for the current device if no device index is passed. 
+
+.. code-block:: cpp
+  
+  CUDAStream getCurrentCUDAStream(DeviceIndex device_index = -1);
+
+.. tip::
+  
+  The current CUDA stream will usually be the default CUDA stream for the device, but it may be different if someone
+  called ``setCurrentCUDAStream`` or used ``StreamGuard`` or ``CUDAStreamGuard``.
+
+
+
+Set CUDA stream
+***************
+
+Pytorch C++ API provides the following ways to set CUDA stream:
+
+1. Set the current stream on the device of the passed in stream to be the passed in stream.
+
+.. code-block:: cpp
+
+  void setCurrentCUDAStream(CUDAStream stream);
+
+.. attention::
+
+  This function may have nosthing to do with the current device. It only changes the current stream on the stream's device.
+  We recommend using ``CUDAStreamGuard``, instead, since it switches to the stream's device and makes it the current stream on that device.
+  ``CUDAStreamGuard`` will also restore the current device and stream when it's destroyed
+
+2. Use ``CUDAStreamGuard`` to switch to a CUDA stream within a scope, it is defined in `CUDAStreamGuard.h`_
+
+.. tip::
+
+  Use ``CUDAMultiStreamGuard`` if you need to set streams on multiple CUDA devices.
+
+CUDA Stream Usage Examples
+**************************
+
+1. Acquiring and setting a CUDA stream
+
+.. code-block:: cpp
+
+  // get a new stream from CUDA stream pool on current device (device 0)
+  // set current stream with this new stream and get current stream
+  torch::cuda::CUDAStream myStream = at::cuda::getStreamFromPool();
+  torch::cuda::setCurrentCUDAStream(myStream);
+  torch::cuda::CUDAStream curStream = torch::cuda::getCurrentCUDAStream();
+  assert(curStream == myStream);
+
+  // get the default CUDA stream on current device (device 0)
+  // set current stream with default stream and get current stream
+  torch::cuda::CUDAStream defaultStream = torch::cuda::getDefaultCUDAStream();
+  torch::cuda::setCurrentCUDAStream(defaultStream);
+  curStream = torch::cuda::getCurrentCUDAStream();
+  assert(curStream == defaultStream);
+  assert(myStream != defaultStream)
+
+.. attention::
+  
+  Above code is running on the same device/gpu (say device 0). `setCurrentCUDAStream` works as expected and set current CUDA stream correctly
```

Comment:
I think this is a little misleading because the above code snippets will work no matter what device you're on.

Reply:
Ah, what I want to say is that this is setCurrentCUDAStream will have strange behavior if passed in stream's device is not the same as current device. Will change a bit.